include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -774,6 +776,77 @@ __group_scan_fits_in_slm(const sycl::queue& __queue, ::std::size_t __n, ::std::s
     return (__n <= __single_group_upper_limit && __max_slm_size >= __req_slm_size);
 }
 
+template <typename _ValueType, typename _UnaryOp>
+struct __gen_transform_input
+{
+    using __out_value_type = typename std::invoke_result<_UnaryOp, _ValueType>::type;
+    template <typename InRng>
+    auto
+    operator()(InRng&& __in_rng, std::size_t __idx) const
+    {
+        return __unary_op(__in_rng[__idx]);
+    }
+    _UnaryOp __unary_op;
+};
+
+struct __simple_write_to_idx
+{
+    template <typename _OutRng, typename ValueType>
+    void
+    operator()(_OutRng&& __out, std::size_t __idx, const ValueType& __v) const
+    {
+        __out[__idx] = __v;
+    }
+};
+
+template <typename _SizeType, typename _Predicate>
+struct __gen_count_pred
+{
+    using __out_value_type = _SizeType;
+    template <typename _InRng>
+    _SizeType
+    operator()(_InRng&& __in_rng, _SizeType __idx)
+    {
+        return __pred(__in_rng[__idx]) ? _SizeType{1} : _SizeType{0};
+    }
+    _Predicate __pred;
+};
+
+template <typename _SizeType, typename _Predicate>
+struct __gen_expand_count_pred
+{
+    template <typename _InRng>
+    auto
+    operator()(_InRng&& __in_rng, _SizeType __idx)
+    {
+        auto ele = __in_rng[__idx];
```

Comment:
Should we forward the `__in_rng` as we use its bracket operator?
I don't think we normally do that, but I suppose we could, right?

Reply:
I think we can, but I am not sure what all it gets us here.

Reply:
Yeah, its probably not any real benefit, and not worth the cost to readability.  Resolving this.